src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs

```diff
@@ -833,6 +833,62 @@ public void SocketSendReceiveBufferSize_SetZero_ThrowsSocketException()
                 Assert.Equal(e.SocketErrorCode, SocketError.InvalidArgument);
             }
         }
+
+        [Fact]
+        public void SendRecvIovMax_Success()
+        {
+            // sending/receiving more than IOV_MAX segments causes EMSGSIZE on some platforms.
+            // This is handled internally so this error shouldn't surface.
+            const int IOV_MAX = 1024; // Linux
+            using (var server = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            {
+                server.BindToAnonymousPort(IPAddress.Loopback);
+                server.Listen(1);
+
+                using (var client = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+                {
+                    client.Connect(server.LocalEndPoint);
+
+                    using (Socket acceptSocket = server.Accept())
+                    {
+                        // client: send data as segmentCount (> IOV_MAX) 1-byte segments.
+                        int segmentCount = IOV_MAX + 1;
+                        var sendBuffer = new byte[segmentCount];
+                        var sendSegments = new List<ArraySegment<byte>>();
+                        for (int i = 0; i < segmentCount; i++)
+                        {
+                            sendBuffer[i] = (byte)(i & 0xff);
+                            sendSegments.Add(new ArraySegment<byte>(sendBuffer, i, 1));
+                        }
+                        SocketError error;
+                        int bytesSent = client.Send(sendSegments, SocketFlags.None, out error);
+
+                        Assert.Equal(segmentCount, bytesSent);
+                        Assert.Equal(SocketError.Success, error);
+
+                        // server: receive data as 1-byte segments.
+                        var receiveBuffer = new byte[segmentCount];
+                        var receiveSegments = new List<ArraySegment<byte>>();
+                        for (int i = 0; i < segmentCount; i++)
+                        {
+                            receiveSegments.Add(new ArraySegment<byte>(receiveBuffer, i, 1));
+                        }
+                        var bytesReceivedTotal = 0;
+                        do
+                        {
+                            int bytesReceived = acceptSocket.Receive(receiveSegments, SocketFlags.None, out error);
+                            receiveSegments.RemoveRange(0, bytesReceived);
```

Comment:
Why is this RemoveRange needed?

Reply:
This is skipping the part we received already. The receives will be truncated at IOV_MAX, so this test will be received in two operations: 1024 segments in the first and 1 segment in the second. Here, here the change is propagating into the user code which is dealing with the fact that TCP can split messages.

Reply:
I understand that it doesn't hurt, but what does it help? Won't the test pass without the RemoveRange?

Reply:
I've increased `segmentCount` to 2400. So there will be 3 calls to `Receive` returning 1024, 1024 and 352 (1-byte) segments. If we don't offset the `receiveSegments`, we will write at offset 0 of `receiveBuffer` for each `Receive`, causing this assert to fail at the end of the test: `Assert.Equal(sendBuffer, receiveBuffer);`.

Reply:
Ah, I didn't see that Assert.Equal.  Makes sense.